File: array/array.py
import random

# initializition of arrays
arr: list[int] = [0] * 5  # [ 0, 0, 0, 0, 0 ]
nums: list[int] = [1, 3, 2, 5, 4]

print("arr:",arr)
print("nums:",nums)

def random_access(nums: list[int]) -> int:
	"""access an element randomly"""
	# get a random index number from range[0, len(nums)-1]
	random_index = random.randint(0, len(nums) - 1)
	# get and return the number at this random index
	random_num = nums[random_index]
	return random_num

def insert(nums: list[int], num: int, index: int):
	"""insert element `num` at `index`(last element will be droped, 
	because arrays are fix-lengthed)"""
	# move elements from index `index` to last one one position backward
	for i in range(len(nums)-1, index , -1):
		nums[i] = nums[i-1]			
	nums[index] = num

def remove(nums: list[int], index: int):
	"""delete the element at index `index`"""
	# move all the elements after `index` one position forward
	for i in range(index, len(nums) - 1):
		nums[i] = nums[i + 1]
	# Reduce the length of the list by 1
	nums.pop()

def traverse(nums: list[int]):
	"""traverse the array"""
	count = 0
	for i in range(len(nums)):
		count += nums[i]
	"""
	# traverse each element directly
	for num in nums:
		count += num
	# or traverse both elements and index
	for i, num in enumerate(nums):
		count += nums[i]
		count += num
	"""
	return count;

def find(nums: list[int], target: int) -> int:
	"""find the provided element in the array"""
	for i in range(len(nums)):
		if nums[i] == target:
			return i
	return -1

def main():
	print("a random number in nums: ", random_access(nums))
	
	insert(nums, 6, 1)
	print("insert 6 in nums at index 1:", nums)
	
	remove(nums, 1)
	print("delete the elememt at index 1:", nums)
	
	print("traverse the array and print the sum of all elements: ", traverse(nums))
	
	print("find 5 in the nums, the index is(-1 if hasn't been found): ", find(nums, 5))
if __name__  == "__main__":
	main()

File: array/array.c
#include <stdio.h>
#include <stdlib.h>

/* access a random element in the array */
int randomAccess(int *nums, int size)
{

    // get a random index in range [0, size)
    int randomIndex = rand() % size;
    // get and return the number at this random index
	int randomNum = nums[randomIndex];
	return randomNum;
}

/* insert element `num` in the index `index` */
void insert(int *nums, int size, int num, int index)
{
	// move the elements including `index` and the elements after it one position backward
	for (int i = size -1; i > index; i--)
		nums[i] = nums[i-1];
	// assign `num` to the element at `index` 
	nums[index] = num;
}

/* delete the element at index `index` */
// the keyword `remove` was occupied by `stdio.h`
void removeItem(int *nums, int size, int index)
{	
	// move the element at `index` and the ones afterward one position forward
	for (int i = index; i < size; i++)
		nums[i] = nums[i + 1];
	
}

int traverse(int *nums, int size)
{
	int count = 0;
	// traverse the array using indice
	for (int i = 0; i < size; i++)
		count += nums[i];
	return count;
}

int main(void)
{
	/* initializition of arrays */
	int arr[5] = {0}; // {0,0,0,0,0}
	int nums[5] = {1, 3, 2, 5, 4};

	/* print initialized arrays out */
	printf("arr: ");
	for (int i = 0; i < 5; i ++) 
		printf("%d ", arr[i]);
	printf("\n");
	
	printf("nums: ");
	for (int i = 0; i < 5; i++)
		printf("%d ",nums[i]);
	printf("\n\n");
	
	/* access a random element in a array */
	for (int i = 0; i < 3; i++)
		printf("a random number in nums: %d\n", randomAccess(nums,5));
	printf("\n");	
	
	/* print the array after insertion */
	printf("insert 6 at index 1 in nums: ");
	insert(nums, 5, 6, 1);
	for (int i = 0; i < 5; i++)
		printf("%d ", nums[i]);
	printf("\n");
	
	/* print the array after deletion */
	printf("delete the element at index 1 in nums: ");
	removeItem(nums, 5, 1);
	for (int i = 0; i < 4; i++ )
		printf("%d ", nums[i]);
	printf("\n");
	
	
	/* print the sum of elements using traverse function */
	printf("the sum of all elements in the array is : %d\n\n", traverse(nums, 4));
	return 0;
}
